napari/_qt/threading.py

```diff
@@ -0,0 +1,563 @@
+
+        - *started*: emitted when the work is started
+        - *finished*: emitted when the work is finished
+        - *yielded*: emitted with yielded values (if generator used)
+        - *returned*: emitted with return value
+        - *errored*: emitted with error object on Exception
+        - *paused*: emitted when a running job has successfully paused
+        - *resumed*: emitted when a paused job has successfully resumed
+        - *aborted*: emitted when a running job is successfully aborted
+
+    If the decorated function is a generator, the (default) returned worker
+    will also provide these methods:
+
+        - *quit*: ask the thread to quit
+        - *toggle_paused*: toggle the running state of the thread.
+        - *send*: send a value into the generator.  (This requires that your
+          decorator function uses the ``value = yield`` syntax)
+
+        # optionally
+        - *start*: if you call your function with ``func(start_thread=False)``,
+          it will also have a `.start()` method that can be used to start
+          execution of the function in another thread.  (useful if you need
+          to connect callbacks to signals prior to execution)
+
+    Parameters
+    ----------
+    func : callable
+        Function to call in another thread.  For communication between threads
+        may be a generator function.
+
+    Returns
+    -------
+    callable
+        function that creates a worker, puts it in a new thread and returns
+        the worker instance.
+
+    Example
+    -------
+
+    .. code-block:: python
+
+        @thread_worker
+        def long_function(start, end):
+            # do setup here.
+            yield # Critical to yield before computation when using generator
+
+            # do work, periodically yielding
+            i = start
+            while i <= end:
+                time.sleep(0.1)
+                yield i
+
+            # do teardown
+            return 'anything'
+
+        # call the function to start running in another thread.
+        worker = long_function(start_thread=False)
+        # connect signals if desired
+        worker.start()
+    """
+
+    def inner_func(func: Callable) -> Callable:
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            return worker_factory(
+                func, *args, worker_class=worker_class, **kwargs
+            )
+
+        return wrapper
+
+    if function is None:
+        return inner_func
+    return inner_func(function)
+
+
+############################################################################
+
+# This is a variant on the above pattern, it uses QThread instead of Qrunnable
+# see https://doc.qt.io/qt-5/threads-technologies.html#comparison-of-solutions
+# It provides more flexibility, but requires that the user manage the threads.
+# With the runnable pattern above, we leverage the QThreadPool.globalInstance()
+# which is wrapped in our own convenience method API.
+
+
+def new_worker_qthread(
```

Comment:
Could play be refactored to use the simpler convenience decorators? 

In other words, when you say this provides more flexibility, what are you actually talking about? Can you give an example in that comment about something you can do with this that you can't do with the decorators?

Reply:
I like going for `QRunnable` and having them in a single `QThreadpool`. Since every workers ends up in a new thread they can access to global context and they subscribe to objects in main thread, via updating objects actually `QRunnable` _can_ get updates from outside context. Are there problems im missing with this approach?

Reply:
I do prefer going with the simpler pattern! So, if I can summarise my vague understanding of your comments:

- for AnimationWorker you used QThreads and QObjects because reasons
- But it *could* be refactored to use these QRunnable workers, it just needs some work that should go in a new PR
- That leaves QThreads and QObjects for people who... Like pain? Or what? I'm still not clear on what the *advantages* would be of keeping this exposed. So I am wondering whether we should just make this thing private and then eventually get rid of it.

Reply:
lol.  Yes, its for people who like pain :). My current understanding of the limitation of using `QRunnable` is that you never get the `QThread` object itself (it's managed by the threadpool).  Take a quick look at the methods on [`QThread`](https://doc.qt.io/qt-5/qthread.html) and you'll see there's a lot more there, like `terminate` and `wait`.  Those are things that might come in handy if we wanted to create our own custom `close_by_hook_or_by_crook` shutdown process... and things that you do _not_ get access to if you allow the QThreadPool to turn your QRunnables into threads.

so:
**QRunnable** = free thread management, but you _must_ build in these little "hooks" to your workers to listen for a "stop this thread" signal.
**QThread** = complete control and responsibility over your threads... much harder to reuse threads (which causes more overhead as you create and destroy them) but easier to destroy rogue threads (at least as far as I've found)

I used QThreads for my animation worker because that was what I was more familiar with at the time.  but yes I _think_ it could be refactored.  I can make this method private right now

Reply:
Very cool, so I suggest let's make this private for now and add that link to QThread methods so people who go fishing in our source are clear on what this provides.